### ISSUE TITLE
ASE Reviewers

### ISSUE THREAD

**Issue by amritbhanu** [ai-se/valueOfsmote#3]
**>>> Summary of discussion leader <<<**

The main contribution of this paper is to show that a careful preprocessing/tuning
(implemented as SMOTUNED) produces dramatically better results in defect
prediction, independently of the classifier used.

During the on-line discussion, the argument was raised that the importance of
preprocessing/tuning has been addressed by several papers in the literature,
including papers from the same authors of this submission.

The on-line discussion also introduced the argument that to actually show that
“Better Data” is Better than “Better Data Miners”, the paper needs to
compare **SMOTUNED with recent defect prediction algorithms that deal with imbalanced
data (see reviewer 2)**: if recent defect prediction algorithms that deal with imbalanced
data outperform SMOTUNED, the suggestion of this paper becomes (1) "Better Data
Miners" is still Better than "Better Data" and (2) practitioners and researchers
are recommended to use SMOTUNED only in the case of using conventional machine
learning techniques on imbalanced datasets.

*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=

1. First reviewer's review:

          **>>> Summary of the submission <<<**

The paper deals with defect prediction and defends the thesis that
preprocessing the data and tuning the parameters is more important than
selecting the right classifies. The authors criticize prior studies aimed at
ranking classifiers based on two arguments: (i) the studies do not evaluate the
classifiers on multiple criteria and (b) they do not investigate how variations
in the data affects the output. To demonstrate their thesis, the authors apply
SMOTUNED, a data preprocessing method aimed reducing class imbalance, and
report dramatic prediction improvements, regardless of the classifier used to
make the predictions.

          **>>> Evaluation <<<**

This is an excellent piece of work. With the growing diffusion of classifiers
to solve software engineering tasks, many researchers have started to look into
the problem: which classifier should be applied to solve a problem/class of
problems? In the last years, numerous ranking studies have been published in
the literature. The present paper shows that these studies are limited as they
ignore the impact of data processing and parameter tuning. As a matter of fact,
the paper shows as a careful preprocessing/tuning produces dramatically better
results in defect prediction, independently of the classifier used. Whilst the
paper focuses on defect predictions, the ideas, arguments and methods can be
exploited to revisit and improve results in other software engineering tasks
that uses classifiers.

The paper has a great potential for impact and will likely generate interest
and discussion at the conference and in the SE community at large.

*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*

2. Second reviewer's review:

          **>>> Summary of the submission <<<**

Defect prediction is a method to predict files that may be defective given
historical data. Features are extracted from source code files and used for the
prediction task. Defect prediction suffers from imbalanced data problem -- most
files are not defective while only a minority is defective. Past studies have
shown that imbalanced data affect the accuracy of defect prediction techniques.
To address the problem with imbalanced data, this paper introduces SMOTUNED
which adaptively tunes SMOTE by using differential evolution and compares
SMOTUNED against SMOTE.

          **>>> Evaluation <<<**

The paper is very well written and organized, I enjoy reading the paper which
flows smoothly. It also shows that SMOTUNED can outperform SMOTE by a big
margin -- improving AUC by 60% and recall by 20%. The experiments have also
been performed on 3,681 Java classes under 5x5 cross-validation study. A nice
literature survey has also been performed on more than 100 defect prediction
papers to investigate their characteristics and limitations.

There are few things that can be improved to strengthen the evaluation and the
literature survey results, and motivate some parts of the work better. I
present my suggestions below.

- [x] 2a. I'm not an expert on SMOTE, but looking at the references [6], it is rather
dated (published in 2002). Machine learning is a fast moving field with
possibly hundreds of papers published from 2002 to 2017 (last 15 years) that
presents techniques to deal with imbalanced data. I'm also not an expert on
defect prediction, but looking at the related work section, it seems there have
been many more recent defect prediction algorithms that deal with imbalanced
data. To quote verbatim: "Menzies et al. [41] undersampled the non-defect class
to balance training data and reported how little information was required to
learn a defect predictor. They found that throwing away data does not degrade
the performance of Naive Bayes and C4.5 decision trees. Other papers show the
usefulness of resampling based on different learners [48], [49], [56]". Thus to
make the paper stronger, SMOTUNED should be compared with: 
    - [x] 2a1. the latest
imbalanced data learning approach in ML literature,
    - [ ] 2a2. recent defect
prediction algorithms that deal with imbalanced data (e.g., the most recent
work among [41], [48], [49], [56] or the ones listed in Table III) -- there
seem to be more works listed in
http://crest.cs.ucl.ac.uk/cow/44/slides/cow44_Wang.pdf (found it through Google
search), 
    - [x] 2a3. latest defect prediction algorithm that does not deal with
imbalanced data (Is it possible that the latest defect prediction algorithm is
immune to imbalanced data? I'm not sure but it can be the demonstrated through
this experiment).

- [ ] 2b. The paper argues of using 4 evaluation metrics. I have several questions: 
    - [ ] 2b1. How
do we know that these 4 are complete? Could there be other important metrics?
Are all 4 necessary, couldn't one or two of them represent the rest? The paper
describes the motivation of using multiple metrics as follows: "management may
be "risk adverse" and hence may elect to maximize recall ... management may be
"cost adverse" and elect to minimize false alarm". Could we just use recall and
false alarm? 
    - [ ] 2b2. Can these two metrics already represent the outcome of the other
metrics? Is it possible that the metrics are highly correlated with one another
(e.g., precision and false alarm)?

- [x] 2c. Out of curiosity, since the title is similar to the problem addressed by this
paper, I checked reference [68] (http://ieeexplore.ieee.org/document/6509481/),
it seems it satisfies all the three criteria. The multiple criteria field of
Table III should be 3 -- The paper considers balance, G-mean, and Area Under
the Curve (AUC). Also, while trying to find reference [26], I found the
following paper that also seems to satisfy all 3 criteria (consider data
imbalanced, evaluated using multiple criteria, and ranked classifiers): Taghi
M. Khoshgoftaar, Kehan Gao: Feature Selection with Imbalanced Data for Software
Defect Prediction. ICMLA 2009: 235-240.

- [x] 2d. I also have a few minor comments. Based on papers that I seen before, table
headings are typically at the top of tables rather than at the bottom (this
stylistic "rule" may not be universal though). The sentence "many researchers
in this area (Wang, Yu et al, Gray ... [70] refer" should be "many researchers
in this area (Wang, Yu ... [70]) refer" - a closing parentheses is missing. 

- [x] 2e. Why
use "oo" and "ck" in the search term? Could we simply use "software" and
"defect prediction"?

+ Well written paper
+ First paper that tunes SMOTE using differential evolution
+ Improve SMOTE by 60% in terms of AUC
+ Survey more than 100 past defect prediction papers and characterize them in
terms of: (1) ranking study, (2) multiple criteria evaluation, and (3) data
imbalance consideration.

- Missing evaluation against recent defect prediction/machine learning
approaches
- [x] 2f.  The survey results may need to be adjusted and the claim may need to be
weakened
- The need for the use of four evaluation metrics is unclear
- The algorithm contribution of the paper is not too much (described in 4
paragraphs of Section 2F)

*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*=--=*

3. Third reviewer's review:

          **>>> Summary of the submission <<<**

This paper empirically evaluates SMOTUNED, which is an auto-tuning version of
SMOTE. SMOTE is a method for addressing the class imbalance problem. This paper
applies SMOTUNED in a 5*5 cross-validation study for 3,681 classes from open
source systems. The result shows that SMOTUNED increases AUC and recall by 60%
and 20%, comparing with No-SMOTE (i.e., do nothing) and SMOTE.

          **>>> Evaluation <<<**

>>> Points for the paper
- The paper is well-structured and well-written
- Table III and Figure 3 are very helpful to understand the overview of related
work
- Experimental settings are well designed
- Results and conclusions are supported statistically
- Recommendation in the conclusion section is insightful

**>>> Points against the paper**
- [x] 3a. Needs to better explain some parts of experimental settings
- [x] 3b. A reproduction package is not available now (due to double blind policy)

>>> Comments
I enjoyed reading this paper and I think the topic has been very well-selected
by the authors. In the following, I provide detailed comments on my issues with
this paper.

- [x] 3c. Page 6 says "In step 3c, we will tell SMOTUNED to optimize for AUC;" I
thought that (1) only AUC is used when tuning parameters of SMOTUNED and (2)
models are tested in according to four evaluation criteria (AUC, precision,
recall, and false alarm). However, 1st paragraph of Section 4.B says "Recall
that when this rig applies SMOTUNED, it optimizes for performance measure M_i
\in {recall, precision, false alarm, AUC}". Does this mean that, in the case of
precision, SMOTUNED is optimized for precision, then models are tested on
precision using the dataset generated by SMOTUNED that is optimized for
precision? Please clarify this point in the paper.

- [x] 3d The caption of Figure 7 describes "for each of these charts, optimize for
performance measure Mi, then test for performance measure Mi." However, I would
like to see that (1) optimize for performance measure "AUC" (or one performance
measure), (2) then test for performance measure M_i. The reason using several
criteria (precision, recall, false alarm, and AUC) is that we should avoid
obtaining a model that is over-fitted on one specific criterion.

- [ ] 3e How do the authors choose the threshold of prediction models?
This paper shows precision, recall, and false alarm, which depend on the
particular threshold used for the classification. Which threshold value is used
in this experiment? Maybe 0.5?
       - What Threshold?

- [x] 3f Could you elaborate a little further on the the sentence "the better function
on line 17 calls a learner on the training data using the proposed new
parameters"? Now we have the training data using the proposed new parameters.
We build a prediction model using a learner (e.g., KNN). 
- [x] 3g. Do we apply the model
to the same training data and obtain the prediction performance for the better
function? or split the training data into n bins, use n - x bins for building a
model and x bins for evaluating the model?

- [x] 3h Figure 5 shows the algorithm of SMOTUNED using DE (differential evolution).
Does rand() method at the line 12 output the range between 0 and 1?

- [x] 3i Section II.D describes how to search scholar.google.com for conjunction of
"software", "defect prediction", "oo", and "ck" published in the last decade.
Why does this paper focus on CK metrics? Process metrics are well-known metrics
in defect prediction and usually provide better performance than product
metrics (static code attribute).

- [x] 3j How do the authors choose 9 datasets? Data comes from the SEACRAFT
repository. When I look at the SEACRAFT repository, especially Keywords, the
number of datasets is 28. Are there any criteria to filter out datasets in the
SEACRAFT repository?

- [x] 3k Minor Comments:
- Figure 1 should be a table, instead of a figure.
- A table caption should be placed above the table.
